### ISSUE TITLE
Slides disappear when deleting recordings

### ISSUE THREAD

**Issue by GladOSkar** [bigbluebutton/bigbluebutton#14172]
**Describe the bug**
After somewhere between 15-60 Minutes of presenting, all slides except the currently viewed one turn entirely white (firefox) or show the "broken image" symbol (chrome).

**To Reproduce**
1. Upload a PDF with 40+ pages to present
2. Go through slides at slow pace (5-10min/slide)

**Expected behavior**
All slides can be clicked through and viewed at all times

**Actual behavior**
At some point, next and previous slides are entirely white, only the last viewed one stays normal.

**Screenshots**
Expected: https://youtu.be/-UxVV9CvFKU (from <https://test.bigbluebutton.org>)
Actual: https://youtu.be/1wCr3x39RT8 (from our servers)

**BBB version:**
`BigBlueButton Server 2.3.15 (2482)`

**Desktop (please complete the following information):**
 - OS: Windows 10, Linux
 - Browser: Firefox + Chrome
 - Version: FF 96.0.1, Chrome 96.0.4664.110
 - Extensions: None

**Additional context**
The error does not appear to occur on <https://test.bigbluebutton.org> - So it might be a config issue? But i haven't changed the config except for upping the slide upload size limit from 30MB to 64MB.

**Comment by ffdixon** [comment]
Interesting.  Is this repeatable or random?  

Could you try using https://demo.bigbluebutton.org/ (which is running 2.4.x) for one of your sessions and let us know if it happens there (if it does, then you can let us know the title and date of the session and we can take a look at the server logs).

**Comment by GladOSkar** [comment]
100% reproducible on our servers

I forgot to mention that this only started to happen after upgrading from 2.2 to 2.3.

I'll try on demo.b... once I can 👍

**Comment by ffdixon** [comment]
We would recommend you upgrade to 2.4.x and see if you can still reproduce it on your servers.

**Comment by GladOSkar** [comment]
I was not able to reproduce on demo. because of the time limit there.

Upgrading our servers to `BigBlueButton Server 2.4.2 (2890)` unfortunately also did not help

**Comment by GladOSkar** [comment]
I'll try finding out how long exactly it takes to trigger

**Comment by GladOSkar** [comment]
Alright I'm back with more details:

The problem appears to occur ~40-50 Minutes after processing/displaying slides (with my test pdf)

The problem occurs because the presentation ressources (`GET https://bbb1.example.com/bigbluebutton/presentation/<id1>/<id1>/<id2>/svg/<pageNumber>`) start 404'ing. The reason one (or rarely a few) pages stay there appears to be browser caching.

Reloading the browser at the session does not help, the slides are fully gone then.

I had a quick look at the logs, but couldn't find anything interesting in `/var/log/bigbluebutton/bbb-web.log` or `/var/log/nginx/bigbluebutton.access.log`.

`nginx/error.log` says:
```
2022/01/24 17:17:09 [error] 1288#1288: *16482 open() "/var/bigbluebutton/<id1>/<id1>/<id2>/svgs/slide<pageNumber>.svg" failed (2: No such file or directory), client: <IP>, server: bbb1.example.com, request: "GET /bigbluebutton/presentation/<id1>/<id1>/<id2>/svg/75 HTTP/2.0", host: "bbb1.example.com", referrer: "https://bbb1.example.com/html5client/join?sessionToken=<sessionToken>"
```

In fact, the entire `/var/bigbluebutton/<id1>` folder has disappeared without the meeting having ended. Do you have any idea what might cause this?

**Comment by GladOSkar** [comment]
Bump

**Comment by GladOSkar** [comment]
> This is a bit confusing for me. Doesn't this mean that the files may have been deleted earlier but due to file caching by the browser, you only saw 404 when the broswer cleared the file from its cache?

No. I already did something similar as you suggested - I change slides every couple of minutes and it always works. During those 40-50 minutes, the log says that the folder+files still exists and the slide changes work (no 404s)

Only after that time the folder disappears entirely and the next slides (the ones that haven't been viewed before) 404 reliably. The previous slides 404 as well since they're usually not cached (depending on how long ago the slide was switched) and teh current slide at the time of failure stays (in cache?) a bit longer.

**Comment by GladOSkar** [comment]
```
17:25 CET Conf start
17:33 CET Slide upload start
17:39 CET Display slides
17:40 CET Last working slide change
18:17 CET Folder gone
```
So in this case about 40-45 minutes depending on where we start counting from.

Running `find /var/bigbluebutton/ -maxdepth 1 -type d -name "*-[0-9]*" -mtime +5` every 4s didn't show anything.

I also set up `auditd` to watch the conference directory in `/var/bigbluebutton` and got this:

```
type=PROCTITLE msg=audit(02/01/2022 18:17:01.665:536) : proctitle=rm -rf /var/bigbluebutton/bd345fd758beb39e1e5012161891a3031a6bc153-1643732723450
type=PATH msg=audit(02/01/2022 18:17:01.665:536) : item=1 name=/var/bigbluebutton/bd345fd758beb39e1e5012161891a3031a6bc153-1643732723450 inode=3408270 dev=08:03 mode=dir,755 ouid=bigbluebutton ogid=bigbluebutton rdev=00:00 nametype=DELETE cap_fp=none cap_fi=none cap_fe=0 cap_fver=0
type=PATH msg=audit(02/01/2022 18:17:01.665:536) : item=0 name=/var/bigbluebutton/ inode=1201667 dev=08:03 mode=dir,755 ouid=bigbluebutton ogid=bigbluebutton rdev=00:00 nametype=PARENT cap_fp=none cap_fi=none cap_fe=0 cap_fver=0
type=CWD msg=audit(02/01/2022 18:17:01.665:536) : cwd=/
type=SYSCALL msg=audit(02/01/2022 18:17:01.665:536) : arch=x86_64 syscall=unlinkat success=yes exit=0 a0=0xffffff9c a1=0x560fe81fd490 a2=0x200 a3=0x560fe81fc010 items=2 ppid=13564 pid=13608 auid=root uid=root gid=root euid=root suid=root fsuid=root egid=root sgid=root fsgid=root tty=(none) ses=247 comm=rm exe=/bin/rm key=(null)
```

So it's definitely something running `rm -rf ...` on the dir, but it's hard to see where from. Next time i'll add a log thing to that `find -exec` so i can see if it might be that

**Comment by GladOSkar** [comment]
```
18:42 Call start
18:43 Slide upload started
18:45 Slides displayed
19:17 Folder gone
```
Curiously a shorter time this time, 32-34 Minutes. Minute `:17` again :thinking:.

The cron job seems not to have triggered.

**Comment by antobinary** [comment]
When I search _carefully_ for `rm -` in bigbluebutton-config/ I see lots of relevant entries in `bbb-record` script.
Could you please comment -- do you handle recording, do you do it on the server (or you offload files to elsewhere), do you have any custom bbb-record or apply-lib.sh functionality that is not copied verbatim from the open source? https://github.com/bigbluebutton/bigbluebutton/blob/v2.4.x-release/bigbluebutton-config/bin

**Comment by GladOSkar** [comment]
We have recordings fully disabled. Which is giving me a hunch right now...

**Comment by GladOSkar** [comment]
Ah yes. I have an hourly cron job running `bbb-record --deleteall`. That's probably it, right? But it's also very much not what i'd have expected that script to do

**Comment by PhMemmel** [comment]
Looks like this causes the problem:

https://github.com/bigbluebutton/bigbluebutton/blob/8404048cfd35280d2dd427fc0c2e93790760c3f7/bigbluebutton-config/bin/bbb-record#L449

It removes the whole directory as `$RAW_PRESENTATION_SRC` is configured to be `/var/bigbluebutton`

**Comment by GladOSkar** [comment]
The `bbb-record` help text actually says
```
--deleteall                        delete all meetings and recordings
```
so this is presumably intended behavior?

But https://docs.bigbluebutton.org/dev/recording.html#delete-all-recordings only mentions it as "Delete all recordings". Some clarification would be nice.

**Comment by antobinary** [comment]
Looks like @danielpetri1 improved the docs in https://github.com/bigbluebutton/bigbluebutton.github.io/pull/372
and @GuiLeme is working on implementing the --deleteall enhancement in #15014 

I will be closing this issue. If anything else remains, let's open a new issue in a narrower area.
